packages/plugins/response-cache-redis/test/response-redis-cache.spec.ts

```diff
@@ -0,0 +1,1155 @@
+    expect(spy).toHaveBeenCalledTimes(1);
+
+    await testInstance.execute(
+      /* GraphQL */ `
+        mutation test($id: ID!) {
+          updateUser(id: $id) {
+            id
+          }
+        }
+      `,
+      {
+        id: 1,
+      }
+    );
+
+    await testInstance.execute(query);
+    expect(spy).toHaveBeenCalledTimes(2);
+  });
+
+  test('should purge cache on demand (typename+id)', async () => {
+    const spy = jest.fn(() => [
+      {
+        id: 1,
+        name: 'User 1',
+        comments: [
+          {
+            id: 1,
+            text: 'Comment 1 of User 1',
+          },
+        ],
+      },
+      {
+        id: 2,
+        name: 'User 2',
+        comments: [
+          {
+            id: 2,
+            text: 'Comment 2 of User 2',
+          },
+        ],
+      },
+    ]);
+    const schema = makeExecutableSchema({
+      typeDefs: /* GraphQL */ `
+        type Query {
+          users: [User!]!
+        }
+
+        type Mutation {
+          updateUser(id: ID!): User!
+        }
+
+        type User {
+          id: ID!
+          name: String!
+          comments: [Comment!]!
+          recentComment: Comment
+        }
+
+        type Comment {
+          id: ID!
+          text: String!
+        }
+      `,
+      resolvers: {
+        Query: {
+          users: spy,
+        },
+        Mutation: {
+          updateUser(_, { id }) {
+            return {
+              id,
+            };
+          },
+        },
+      },
+    });
+
+    const testInstance = createTestkit([useResponseCache({ cache })], schema);
+
+    const query = /* GraphQL */ `
+      query test {
+        users {
+          id
+          name
+          comments {
+            id
+            text
+          }
+        }
+      }
+    `;
+
+    // query and cache
+    await testInstance.execute(query);
+    // get from cache
+    await testInstance.execute(query);
+    // so queried just once
+    expect(spy).toHaveBeenCalledTimes(1);
+
+    expect(await redis.exists('User:1')).toBeTruthy();
+    expect(await redis.exists('User:2')).toBeTruthy();
+    expect(await redis.exists('User:3')).toBeFalsy();
+
+    expect(await redis.exists('Comment:1')).toBeTruthy();
+    expect(await redis.exists('Comment:2')).toBeTruthy();
+    expect(await redis.exists('User:3')).toBeFalsy();
+
+    const commentMembers = await redis.smembers('Comment');
+    expect(commentMembers).toHaveLength(1);
+
+    const comment1Members = await redis.smembers('Comment:1');
+    expect(comment1Members).toHaveLength(1);
+
+    const comment2Members = await redis.smembers('Comment:2');
+    expect(comment2Members).toHaveLength(1);
+
+    await cache.invalidate([{ typename: 'Comment', id: 2 }]);
+
+    const commentMembersInvalidated = await redis.smembers('Comment');
+    expect(commentMembersInvalidated).toHaveLength(1);
+
+    const comment1MembersInvalidated = await redis.get('Comment:1');
+    expect(comment1Members).toHaveLength(1);
```

Comment:
It seems like this assertion should be `expect(comment1MembersInvalidated).toHaveLength(1);`?

Reply:
Yes, you are correct. I actually I found a bug  (a get vs smembers) in that test case and have corrected as well as added code comments to make it a little more clear what is happening and being verified.